### ISSUE TITLE
Neet to fetch metadata after source and use them directly bevore output

### ISSUE THREAD

**Issue by PB-SF** [savonet/liquidsoap#1450]
Liquidsoap 1.4.2
Ubuntu 18.04

In my script I use the StereoTool integrated with the `pipe` parameter.
`s = pipe(restart_on_error=true, process='/path/to/stereo_tool - - -s /path/to/stereotool.sts -q', drop_metadata(s))`

This only works smoothly if I discard the metadata in the `pipe` parameter with `drop_metadata`.

How can I fetch the metadata directly after the source in order to insert it again later in the script (ideally directly before` output`)?

thanks in advance

**Comment by toots** [comment]
Hi! Could you explain more in details what you are currently doing and what you are expecting to be able to do? Some script examples would help. Thanks!

**Comment by PB-SF** [comment]
Hi @toots  - i try as best as possible, my english is not the best :)

I have a script in which several playlists are stored. Plus a harbor for live moderation.

StereoTool takes care of the final sound processing - so far separately in each individual studio. That's why I would like to integrate the StereoTool into Liquidsoap directly in front of the Icecast server. It works so far - but only if I include `drop_metadata` in the `pipe` parameter. Without this `drop_metadata`, the stereo tool stops with every transition to a new title and does not restart properly. Only choppy music can then be heard on the stream.

With `drop_metadata` in `pipe` i have no metadata displayed in icecast and also not in the listeners player.

As there is no solution in sight to this problem with StereoTool, I try to work around it in the Liquidsoap Script.

It currently looks like this in the important parts:

```
#PLAYLISTS
playlist_mixed = playlist(id="playlist_mixed", mode="random", reload_mode="watch", conservative=true, default_duration=10., length=20., "/path/to/m3u")
playlist_mixed = audio_to_stereo(id="stereo_playlist_mixed", playlist_mixed)
playlist_mixed = cue_cut(id="cue_playlist_mixed", playlist_mixed)

playlist_top = playlist(id="playlist_top", mode="random", reload_mode="watch", conservative=true, default_duration=10., length=20., "/path/to/m3u")
playlist_top = audio_to_stereo(id="stereo_playlist_top", playlist_top)
playlist_top = cue_cut(id="cue_playlist_top", playlist_top)

#MIX THE PLAYLISTS
radio = random(id="standard_playlists", weights=[5, 1], [playlist_mixed, playlist_top])

#CROSSFADE
radio = crossfade(smart=true, duration=4.50, fade_out=3.00, fade_in=3.00, radio)

#LIVE HARBOR
live = audio_to_stereo(input.harbor("/live", id="live_harbor", port=8099, password="hackme", icy=true, icy_metadata_charset="UTF-8", metadata_charset="UTF-8", buffer=5., max=10.))
ignore(output.dummy(live, fallible=true))

#LIVE FALLBACK
radio = fallback(id="live_fallback", replay_metadata=false, track_sensitive=false, [live, radio])

#STEREOTOOL (WORK ONLY WITH drop_metadata)
radio = pipe(restart_on_error=true, process='/path/to/stereo_tool_cmd_64 - - -s /path/to/stereotool.sts -q', drop_metadata (radio))
radio = mksafe(radio)

# OUTPUT TO ICECAST
```

In order to get stereotool run continuously i would like to intercept the metadata right after the #LIVE FALLBACK and then insert it again after the #STEREOTOOL and before the #OUTPUT TO ICECAST - if that is possible ?!

Maybe there is another possibility?

Many greetings and thanks

**Comment by toots** [comment]
Hi! Yes, there is another possibility! Just to explain the context: the process used in the `pipe` operator has its own internal buffer so, when we want to apply metadata to the output, it's impossible for us to know in advance when the beginning of the corresponding track will come out of the process.

Thus, our first solution is to totally shutdown the process, flush its buffer, apply the metadata, and restart it. This allows for an exact solution but does not work with some program such as StereoTool.

For this program, we have implemented another solution, which is to apply a fixed delay before applying metadata to the output. Usually, for a given fixed piping program, you should be able to find the delay that is appropriate to re-inject the metadata at the right time.

This mode is activated using the `replay_delay`:
> Replay track marks and metadata from the input source on the output
> after a given delay. If negative (default) close and flush the process
> on each track and metadata to get an exact timing.

Hope this helps! I'm closing this ticket now, feel free to re-open or follow-up on slack. Thks!

**Comment by PB-SF** [comment]
Hi @toots 

Sorry that I have to open this thread again.

I've tested a lot in the last few days. The pipe process runs well on the outside stream.

Unfortunately I now have the problem that my log files are flooded.

In the Liquidsoap Log I get the following entry several times per second:

![2021-01-19_10h17_35](https://user-images.githubusercontent.com/45131983/105013792-e5d03880-5a3f-11eb-9386-eba31b5a7a91.png)

Here I have no problem with it, because I set the log level to 5 for the test. In the productive area, I set it back to standard 3.

But in the Shoutcast Log I also get a metadata update several times per second. This is where it becomes problematic because I need these log files for later analysis.

![2021-01-19_10h18_15](https://user-images.githubusercontent.com/45131983/105013822-ee287380-5a3f-11eb-8f63-8b5feeefbf23.png)

That makes the log files huge in a short time.

Without the pipe I don't have the problem.

Is there also a solution for this?

Thank you and best regards.
